mirage/client/dns_client_mirage.ml

```diff
@@ -192,10 +192,10 @@ The format of a nameserver is:
       Lwt_condition.wait t.timer_condition >>= fun () ->
       loop ()
 
-    let read_udp t port ip ~src ~dst:_ ~src_port data =
+    let read_udp t port ip ~src ~dst:_ ~src_port:_ data =
       (* TODO: compare dst being us *)
-      if (port = src_port && Ipaddr.compare ip src = 0) ||
-         (src_port = t.last_udp_port && Ipaddr.(compare ip (V4 V4.any) = 0)) &&
+      if (Ipaddr.compare ip src = 0 ||
+          (port = t.last_udp_port && Ipaddr.(compare ip (V4 V4.any) = 0))) &&
```

Comment:
In all call sites, we pass `dst_port` to `read_udp` as `port` when setting up the listener. So the `port = dst_port` should be unnecessary.
I'm not sure about comparing `port` with `t.last_udp_port`. In `create` we update `t` with `last_udp_port` to the generated port. In the other callsite (`send_recv`) we *don't* update `last_udp_port`. With that in mind I'm confused what it is we want with `last_udp_port` which to me seems to be in fact the first udp port. It also seems to me that we will never receive answers unless we happen to send and receive on the port generated in `create`.

Reply:
I am now even more confused with the `V4.any` bit. If `generate_udp_port` fails to generate a port and reuses `last_udp_port` we will be overriding the `V4.any` listener in `send_recv`. I'm not sure what the purpose is in setting up the listener in `create` then.

Reply:
Sorry for the confusion. The reasoning is as follows:
- the resource of UDP porrts is limited to 2^16 - 1024 (since DNS servers dislike being asked from a privileged (<1024) port)
- we would like to have as much entropy as possible to prevent DNS spoofing (https://en.wikipedia.org/wiki/DNS_spoofing), namely: a random DNS identifier, random UDP source port
- we may end up in a situation that all UDP ports have been used (when we have 2^16 - 1024 requests concurrently)

In general, we allocate a not-yet used UDP port in `send_recv`, listen (with `read_udp`) on it, and unlisten once a reply or timeout happened.

Maybe the `last_udp_port` is over-engineered, but the idea is: if there's no free UDP port, use a port reserved at startup (in `create`) to be able to send requests and receive replies. Since I wrote this, I think this is overly complex and when we're out of UDP ports we're in trouble anyways. I'll push a subsequent commit that removes this complexity.

Reply:
Hi, do you think this reasoning  is also relevant for qubes-mirage-firewall?

Reply:
In qubes-mirage-firewall we only ever talk to the qubesdb-provided dns resolver, and use by default a random UDP port (but still have a `last_resort_port` allocated at start). I don't think we need to change anything in qubes-mirage-firewall (though we could remove the `last_resort_port` and fail if there's no free UDP port -- the downside is that DNS resolution may be required for rule matching, and the qubes-mirage-firewall does NAT also for UDP, thus more UDP ports are used).